piker/clearing/_ems.py

```diff
@@ -347,221 +287,307 @@ async def process_broker_trades(
         # https://www.python.org/dev/peps/pep-0636/#going-to-the-cloud-mappings
         if action in ('cancel',):
 
             # check for live-broker order
-            if brid:
+            if live_entry:
+
+                msg = BrokerdCancel(
+                    oid=oid,
+                    reqid=reqid or live_entry.reqid,
+                    time_ns=time.time_ns(),
+                )
+
+                # send cancel to brokerd immediately!
                 log.info("Submitting cancel for live order")
-                await client.submit_cancel(reqid=brid)
 
-            # check for EMS active exec
+                # NOTE: cancel response will be relayed back in messages
+                # from corresponding broker
+                await brokerd_order_stream.send(msg.dict())
+
             else:
+                # might be a cancel for order that hasn't been acked yet
+                # by brokerd so register a cancel for then the order
+                # does show up later
+                dark_book._ems_entries[oid] = msg
+
+                # check for EMS active exec
                 try:
+                    # remove from dark book clearing
                     dark_book.orders[symbol].pop(oid, None)
 
-                    await ctx.send_yield({
-                        'resp': 'dark_cancelled',
-                        'oid': oid
-                    })
+                    # tell client side that we've cancelled the
+                    # dark-trigger order
+                    await client_order_stream.send(
+                        Status(
+                            resp='dark_cancelled',
+                            oid=oid,
+                            time_ns=time.time_ns(),
+                        ).dict()
+                    )
+
                 except KeyError:
                     log.exception(f'No dark order for {symbol}?')
 
+        # TODO: 3.10 struct-pattern matching and unpacking here
         elif action in ('alert', 'buy', 'sell',):
 
-            sym = cmd['symbol']
-            trigger_price = cmd['price']
-            size = cmd['size']
-            brokers = cmd['brokers']
-            exec_mode = cmd['exec_mode']
+            msg = Order(**cmd)
 
-            broker = brokers[0]
+            sym = msg.symbol
+            trigger_price = msg.price
+            size = msg.size
+            exec_mode = msg.exec_mode
+            broker = msg.brokers[0]
 
             if exec_mode == 'live' and action in ('buy', 'sell',):
 
-                # register broker id for ems id
-                order_id = await client.submit_limit(
+                if live_entry is not None:
 
+                    # sanity check on emsd id
+                    assert live_entry.oid == oid
+
+                    # if we already had a broker order id then
+                    # this is likely an order update commmand.
+                    log.info(f"Modifying order: {live_entry.reqid}")
```

Comment:
probably should log the backend name here and that it's modifying and existing *live* `brokerd` order.